apisix/plugins/forward-auth.lua

```diff
@@ -0,0 +1,145 @@
+local http   = require("resty.http")
+
+local schema = {
+    type = "object",
+    properties = {
+        host = {type = "string"},
+        ssl_verify = {
+            type = "boolean",
+            default = true,
+        },
+        request_headers = {
+            type = "array",
+            default = {},
+            items = {type = "string"},
+            description = "client request header that will be sent to the authorization"
+        },
+        upstream_headers = {
+            type = "array",
+            default = {},
+            items = {type = "string"},
+            description = "authorization response header that will be sent to the upstream"
+        },
+        client_headers = {
+            type = "array",
+            default = {},
+            items = {type = "string"},
+            description = "authorization response header that will be sent to"
+                           .. "the client when authorize failure"
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            maximum = 60000,
+            default = 3000,
+            description = "timeout in milliseconds",
+        },
+        keepalive = {type = "boolean", default = true},
+        keepalive_timeout = {type = "integer", minimum = 1000, default = 60000},
+        keepalive_pool = {type = "integer", minimum = 1, default = 5},
+    },
+    required = {"host"}
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 2002,
+    name = "forward-auth",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.access(conf, ctx)
+    local auth_headers = {
+        ["X-Forwarded-Proto"] = core.request.get_scheme(ctx),
+        ["X-Forwarded-Method"] = core.request.get_method(),
+        ["X-Forwarded-Host"] = core.request.get_host(ctx),
+        ["X-Forwarded-Uri"] = core.request.get_uri(ctx),
+        ["X-Forwarded-For"] = core.request.get_remote_client_ip(ctx),
+    }
+
+    -- append headers that need to be get from the client request header
+    if #conf.request_headers > 0 then
+        for _, header in ipairs(conf.request_headers) do
+            auth_headers[header] = core.request.header(ctx, header)
+        end
+    else
+        auth_headers = core.table.merge(core.request.headers(), auth_headers)
+    end
+
+    local params = {
+        method = "GET",
+        headers = auth_headers,
+        keepalive = conf.keepalive,
+        ssl_verify = conf.ssl_verify
+    }
+
+    if conf.keepalive then
+        params.keepalive_timeout = conf.keepalive_timeout
+        params.keepalive_pool = conf.keepalive_pool
+    end
+
+    local httpc = http.new()
+    httpc:set_timeout(conf.timeout)
+
+    local res, err = httpc:request_uri(conf.host, params)
+
+    -- block by default when authorization service is unavailable
+    if not res or err then
+        core.log.error("failed to process forward auth, err: ", err)
+        return 403
+    end
```

Comment:
```suggestion
    if not res then
        core.log.error("failed to process forward auth, err: ", err)
        return 403
    end
```
Would it be better to judge this way?

Reply:
I don't understand how to do it really right, I noticed that in other plugin implementations there are those that use `not res` alone, those that use `err` alone and those that use `not res or err`, which one should we take?

https://github.com/apache/apisix/blob/58fec8f5eb599c93c3491e200a4ab66907a3dd24/apisix/plugins/authz-keycloak.lua#L251-L254

https://github.com/apache/apisix/blob/58fec8f5eb599c93c3491e200a4ab66907a3dd24/apisix/plugins/serverless/generic-upstream.lua#L110-L113

https://github.com/apache/apisix/blob/58fec8f5eb599c93c3491e200a4ab66907a3dd24/apisix/plugins/wolf-rbac.lua#L145-L147

I hope to get your guidance.
cc @spacewander

Reply:
Then, referring to the way it is usually used, I will determine if `err` is `not nil`.

```lua
if err then
    xxx
end
```

Reply:
changed

Reply:
Also, I think there needs to be a specification of what rules developers should follow to handle error checking (at least on the http client), the mix of different usages is confusing.

Reply:
Currently, we have to enforce it with code review. Maybe we can invest in linter in the future.

Reply:
I think it should be used

````lua
if res then
     err
end
````

instead of

```lua
if err then
     err
end
```

this way of writing may be more suitable for languages such as `Go`

the common exception function return in OpenResty is `return nil, err`, the first value is enough to know whether to handle `err`, and you can refer to more code and test cases in `lua-resty-core`, such as: https://github.com/openresty/lua-resty-core/blob/e5217414669c100b334940b250d5340911a02dd0/t/ctx.t#L133-L143

Reply:
changed